dspace-server-webapp/src/test/java/org/dspace/app/rest/PrimaryBitstreamControllerIT.java

```diff
@@ -0,0 +1,296 @@
+        bundle.setPrimaryBitstreamID(bitstream);
+        Bundle bundle2 = BundleBuilder.createBundle(context, item).withName("Bundle2").build();
+        Bitstream bitstream2 = createSecondBitstream(bundle2);
+        context.restoreAuthSystemState();
+
+        String token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(put(getBundleUrl(bundle.getID()))
+                                     .contentType(textUriContentType)
+                                     .content(getBitstreamUrl(bitstream2.getID())))
+                        .andExpect(status().isUnprocessableEntity());
+        // verify primaryBitstream is still the original one
+        bundle = context.reloadEntity(bundle);
+        Assert.assertEquals(bitstream, bundle.getPrimaryBitstream());
+    }
+
+    @Test
+    public void testDeletePrimaryBitstream() throws Exception {
+        bundle.setPrimaryBitstreamID(bitstream);
+
+        String token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(delete(getBundleUrl(bundle.getID())))
+                        .andExpect(status().isNoContent());
+        // verify primaryBitstream was actually deleted
+        bundle = context.reloadEntity(bundle);
+        Assert.assertNull(bundle.getPrimaryBitstream());
+        // verify bitstream itself still exists
+        Assert.assertEquals(1, bundle.getBitstreams().size());
+        Assert.assertEquals(bitstream, bundle.getBitstreams().get(0));
+    }
+
+    @Test
+    public void testDeletePrimaryBitstreamBundleNotFound() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(delete(getBundleUrl(UUID.randomUUID())))
+                        .andExpect(status().isNotFound());
+    }
+
+    @Test
+    public void testDeletePrimaryBitstreamBundleNonExisting() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(delete(getBundleUrl(bundle.getID())))
+                        .andExpect(status().isBadRequest());
+        // verify primaryBitstream is still null
+        bundle = context.reloadEntity(bundle);
+        Assert.assertNull(bundle.getPrimaryBitstream());
+    }
+
+    private String getBundleUrl(UUID uuid) {
+        return "/api/core/bundles/" + uuid + "/primaryBitstream";
+    }
+
+    private String getBitstreamUrl(UUID uuid) {
+        return "/api/core/bitstreams/" + uuid;
+    }
+
+    private Bitstream createSecondBitstream(Bundle bundle) throws Exception {
+        String bitstreamContent = "Second Bitstream";
+        try (InputStream is = IOUtils.toInputStream(bitstreamContent, CharEncoding.UTF_8)) {
+             return BitstreamBuilder.createBitstream(context, bundle, is)
+                                         .withName("Bitstream2")
+                                         .withMimeType("text/plain")
+                                         .build();
+        }
+    }
+}
```

Comment:
These tests look good. But, we appear to be missing tests for non-Admin users.  I'd expect tests for the following scenarios:

* Tests that prove that a Community or Collection Admin can set a primary bitstream **if the Item belongs in their Community or Collection**
* Tests which prove that Community or Collection Admin cannot set the primary bitstream for an Item that is **not** in their Community or Collection.
* Tests which prove that a user with Item Admin privileges can also set the Primary bitstream of the Item they have admin privileges on.
* Tests which prove that a normal user (no special permissions) or anonymous user cannot modify primary bitstream of an Item.